pennylane/transforms/optimization/optimization_utils.py

```diff
@@ -0,0 +1,162 @@
+from pennylane.wires import Wires
+
+
+def _find_next_gate(wires, op_list):
+    """Given a list of operations, finds the next operation that acts on at least one of
+    the same set of wires, if present.
+
+    Args:
+        wires (Wires): A set of wires acted on by a quantum operation.
+        op_list (list[Operation]): A list of operations that are implemented after the
+            operation that acts on ``wires``.
+
+    Returns:
+        int or None: The index, in `op_list`, of the earliest gate that uses one or more
+            of the same wires, or None if no such gate is present.
+    """
+    next_gate_idx = None
+
+    for op_idx, op in enumerate(op_list):
+        if len(Wires.shared_wires([wires, op.wires])) > 0:
+            next_gate_idx = op_idx
+            break
+
+    return next_gate_idx
+
+
+def _yzy_to_zyz(middle_yzy):
+    """Converts a set of angles representing a sequence of rotations RY, RZ, RY into
+    an equivalent sequence of the form RZ, RY, RZ.
+
+    Any rotation in 3-dimensional space (or, equivalently, any single-qubit unitary)
+    can be expressed as a sequence of rotations about 3 axes in 12 different ways.
+    Typically, the arbitrary single-qubit rotation is expressed as RZ(a) RY(b) RZ(c),
+    but there are some situations, e.g., composing two such rotations, where we need
+    to convert between representations. This function converts the angles of a sequence
+
+    .. math::
+
+       RY(y_1) RZ(z) RY(y_2)
+
+    into the form
+
+    .. math::
+
+       RZ(z_1) RY(y) RZ(z_2)
+
+    This is accomplished by first converting the rotation to quaternion form, and then
+    extracting the desired set of angles.
+
+    Args:
+        y1 (float): The angle of the first ``RY`` rotation.
+        z (float): The angle of the inner ``RZ`` rotation.
+        y2 (float): The angle of the second ``RY`` rotation.
+
+    Returns:
+        [float, float, float]: A list of rotation angles in the ZYZ representation.
+    """
+    if allclose(stack(middle_yzy), cast_like(zeros(3), stack(middle_yzy))):
+        return stack([0.0, 0.0, 0.0])
+
+    y1, z, y2 = middle_yzy[0], middle_yzy[1], middle_yzy[2]
+
+    # First, compute the quaternion representation
+    # https://ntrs.nasa.gov/api/citations/19770024290/downloads/19770024290.pdf
+    qw = cos(z / 2) * cos(0.5 * (y1 + y2))
+    qx = sin(z / 2) * sin(0.5 * (y1 - y2))
+    qy = cos(z / 2) * sin(0.5 * (y1 + y2))
+    qz = sin(z / 2) * cos(0.5 * (y1 - y2))
+
+    # Now convert from YZY Euler angles to ZYZ angles
+    # Source: http://bediyap.com/programming/convert-quaternion-to-euler-rotations/
+    z1_arg1 = 2 * (qy * qz - qw * qx)
+    z1_arg2 = 2 * (qx * qz + qw * qy)
+    z1 = arctan2(z1_arg1, z1_arg2)
+
+    y = arccos(qw ** 2 - qx ** 2 - qy ** 2 + qz ** 2)
+
+    z2_arg1 = 2 * (qy * qz + qw * qx)
+    z2_arg2 = -2 * (qx * qz - qw * qy)
+    z2 = arctan2(z2_arg1, z2_arg2)
+
+    return stack([z1, y, z2])
+
+
+def _fuse_rot_angles(angles_1, angles_2):
```

Comment:
Feel free to turn this off using a pylint comment (or, alternatively, it might be a sign it could be factored out into smaller unit functions)